store/errors.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package store
+
+import (
+	"fmt"
+)
+
+// ErrInvalidInput indicates an error that has occured due to an invalid input.
+type ErrInvalidInput struct {
+	Entity string      // The entity which was sent as the input.
+	Field  string      // The field of the entity which was invalid.
+	Value  interface{} // The actual value of the field.
+}
+
+func NewErrInvalidInput(entity, field string, value interface{}) *ErrInvalidInput {
+	return &ErrInvalidInput{
+		Entity: entity,
+		Field:  field,
+		Value:  value,
+	}
+}
+
+func (e *ErrInvalidInput) Error() string {
+	return fmt.Sprintf("invalid input: entity: %s field: %s value: %s", e.Entity, e.Field, e.Value)
+}
+
+// ErrLimitExceeded indicates an error that has occured because some value exceeded a limit.
+type ErrLimitExceeded struct {
+	What  string // What was the object that exceeded.
+	Count int    // The value of the object.
+	meta  string // Any additional metadata.
+}
+
+func NewErrLimitExceeded(what string, count int, meta string) *ErrLimitExceeded {
+	return &ErrLimitExceeded{
+		What:  what,
+		Count: count,
+		meta:  meta,
+	}
+}
+
+func (e *ErrLimitExceeded) Error() string {
+	return fmt.Sprintf("limit exceeded: what: %s count: %d metadata: %s", e.What, e.Count, e.meta)
+}
+
+// ErrConflict indicates a conflict that occured.
```

Comment:
What's the reason behind `ErrConflict` vs, say, `RecordExists`? What kind of other conflicts type be used for?

Reply:
Right now, it's very hard to say. Because this is just the first PR. That's why I kept it generic for now. But 0/5 on changing it right now.